WordPress/gutenberg
# `UnitControl`: tidy up utils and types

## Description


_While working on #38753 (specifically on the type of the `value` prop of the `NumberControl` component), it became clear that we should first untangle a few components using `NumberControl` — and `UnitControl` is one of them._

This PR tries to refactor and tidy up some of the types (and consequently the logic) of the `UnitControl`, in particular its many utility functions. **With that in mind, this PR's main goal is to add a clear internal split between the numeric quantity and the unit parts that together compose the value held by `UnitControl`**, all while introducing the least amount of runtime change possible and ideally no changes to the public API of `UnitControl` (despite it being flagged as an experimental component).

### Changes overview

- in order to avoid confusion around the meaning of "value", this PR renamed (and adjusted types) for a few variables/functions:
    - the value passed to `UnitControl`, which internally goes through the component's parsing logic, is now referred to as _"Raw Value"_
    - the numeric value parsed by the component's internal logic is now referred to as _"quantity",_ and is always assumed to be a `number` (when defined) 
    - the unit value parsed by the component's internal logic is now referred to as _"unit"_, and is always assumed to be a `string` (when defined) 
    - to recap: a _"raw value"_  passed to `UnitControl` (e.g `14`, `'2.5rem'`, `'600%'`...) is parsed by the internal logic and split into a _"quantity"_ (of type `number`) and a _"unit"_ (of type `string`)
    - some utility functions have also been renamed to be more explicit (and often more verbose), and their JS docs have been updated. This will hopefully add more clarity to what each utility function is supposed to do.
 - the `default` property of the `WPUnitControlUnit` has changed to be a `number`, following the type of a parsed  "quantity" (it was also already documented as such in the README)
- the type alias `Value` has been removed, and replaced by more explicit `string` and/or `number` types where necessary
- the type alias `WPUnitControlUnitList` has been removed, and has been replaced by `WPUnitControlUnit[]` (which means that those values can only be array, and can't be `false` anymore)
- the `value` prop of `UnitControl` was previously typed as required, this PR changes it to be non-required (it was already being omitted or passed as `null` in the repo, and advertised as such in the README anyway)
- the native `UnitControl` behaviour has slightly changed around how a value of `[]` for the `units` prop is handled (see [code changes](https://github.com/WordPress/gutenberg/pull/38987/files#diff-5eefbe2941ac9aaafe134fccc95d9b1a348058caee56abc658967e636c3f44ed)) 

### List of introduced runtime changes

- As a consequence of the simplification and removal of the `WPUnitControlUnitList` type, when passing a list of units,  all `false` values were replaced by `[]`
- When passing values to `UnitControl`, changed values from `null` to `undefined`
- Since the parsed _"quantity"_ is considered a `number`, the parsing logic internal to `UnitControl` now returns `undefined` if a quantity can't be parsed (it previously returned `''`). As a consequence, the fallback to `''` has been moved to each consumer the parsing utility (if necessary).
- The `default` value of all "base" units has changed from `''` to `undefined`
- As a consequence of the fact that `value` can now be `undefined` (and that it is already passed as `null` or `undefined` anyway across the repo), a few more checks and fallbacks have been added to the `UnitControl` component and some of its utility functions

## Testing Instructions





- Project builds
- All tests pass
- Storybook stories work as expected
- Components using `UnitControl` work as expected, both in Storybook and in the editor:
  - `FontSizePicker`
  - `BoxControl`
  - `FocalPointPicker`
  - `ToolsPanel`
- Components using `UnitControl` in the `@wordpress/block-editor` package work as expected:
  - `BorderRadiusControl`
  - `LetterSpacingControl`
  - `LineHeightControl`
  - `BorderWidthEdit`
  - `GapEdit`
  - "Flow" layout inspector controls
  - Column block width control
  - Cover block dimension controls
  - Post featured Image's dimension controls
  - Search block dimensions
  - Spacer block dimensions controls
- Components using `UnitControl` in the `@wordpress/edit-site` package work as expected:
  - Controls in the border panel and dimensions panel

## Types of changes





Refactor

## Checklist:
- N/A I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). 
- N/A I've updated related schemas if appropriate.
